### PR TITLE
Add unstable `-Z direct-access-external-data` cmdline flag for `rustc`

### LINKED ISSUE #118053: [loongarch*-linux] `-Crelocation-model=static` should not access external data directly
I tried this code:

```rust
use std::ffi::CStr;
use std::os::raw::c_char;

extern "C" {
    static environ: *const *const c_char;
}

fn main() {
    unsafe {
        let mut env_ptr = environ;

        while !(*env_ptr).is_null() {
            let c_str_ptr = *env_ptr;
            let rust_str = CStr::from_ptr(c_str_ptr);

            println!("{:?}", rust_str.to_string_lossy());
            env_ptr = env_ptr.offset(1);
        }
    }
}
```

I expected to see the environments variables printed.

Instead, this happened: `Aborted (core dumped)`

### Meta

`rustc --version --verbose`:
```
rustc 1.74.0 (79e9716c9 2023-11-13)
binary: rustc
commit-hash: 79e9716c980570bfd1f666e3b16ac583f0168962
commit-date: 2023-11-13
host: loongarch64-unknown-linux-gnu
release: 1.74.0
LLVM version: 17.0.4
```

<details><summary>Backtrace</summary>
<p>

```
thread 'main' panicked at t.rs:12:16:
misaligned pointer dereference: address must be a multiple of 0x8 but is 0x28ee81ef1c000aef
stack backtrace:
   0: rust_begin_unwind
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:597:5
   1: core::panicking::panic_nounwind_fmt
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:106:14
   2: core::panicking::panic_misaligned_pointer_dereference
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:193:5
   3: t::main
             at ./t.rs:12:16
   4: core::ops::function::FnOnce::call_once
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread caused non-unwinding panic. aborting.
Aborted (core dumped)
```

</p>
</details>

### PR DESCRIPTION
**Pull request by heiher**
The new flag has been described in the Major Change Proposal at https://github.com/rust-lang/compiler-team/issues/707

Fixes #118053

### CODE REVIEW

**Comment by heiher** [comment]
cc @xry111

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/compiler-team/issues/707.
@rustbot blocked

File: tests/codegen/direct-access-external-data.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// [INDIRECT] compile-flags: -C relocation-model=static -Z direct-access-external-data=no
+
+#![crate_type = "rlib"]
+
+// DEFAULT: @var = external {{.*}} global i32
+// DIRECT: @var = external dso_local {{.*}} global i32
+// INDIRECT: @var = external {{.*}} global i32
+
+extern "C" {
+    static var: i32;
```
Comment (oli-obk) [inline tests/codegen/direct-access-external-data.rs:15]:
```suggestion
    static VAR: i32;
```

Comment (heiher) [reply tests/codegen/direct-access-external-data.rs]: Thanks.

**Comment by heiher** [comment]
https://github.com/rust-lang/compiler-team/issues/707 done.
@rustbot ready

**Pull request merged**